smart-contract-extensions/contracts/KeyPurchaser.sol

```diff
@@ -0,0 +1,141 @@
+ * the subscription but it does not. This should be handled by the frontend.
+ */
+contract KeyPurchaser is Initializable, Stoppable
+{
+  using Address for address payable;
+  using SafeERC20 for IERC20;
+
+  // set on initialize and cannot change
+  IPublicLock public lock;
+  uint public maxKeyPrice;
+  uint public maxPurchaseCount;
+  uint public renewWindow;
+  uint public renewMinFrequency;
+  bool public isSubscription;
+
+  // admin can change these anytime
+  string public name;
+  bool internal disabled;
+
+  // store minimal history
+  mapping(address => uint) public timeOfLastPurchase;
+
+  /**
+   * @notice Called once to set terms that cannot change later on.
+   */
+  function initialize(
+    IPublicLock _lock,
+    uint _maxKeyPrice,
+    uint _renewWindow,
+    uint _renewMinFrequency,
+    bool _isSubscription
+  ) public
+    initializer()
+  {
+    require(_maxKeyPrice > 0, 'INVALID_MAX_KEYPRICE');
+    require(_renewWindow > 0 || !_isSubscription, 'INVALID_RENEW_WINDOW');
+    require(_renewMinFrequency > 0 || !_isSubscription, 'INVALID_RENEW_MINFREQUENCY');
+
+    _initializeAdminRole();
+    lock = _lock;
+    maxKeyPrice = _maxKeyPrice;
+    renewWindow = _renewWindow;
+    renewMinFrequency = _renewMinFrequency;
+    isSubscription = _isSubscription;
+  }
+
+  /**
+   * @notice Used by admins to update metadata which may be leveraged by the frontend.
+   */
+  function config(
+    string memory _name,
+    bool _disabled
+  ) public
+    onlyAdmin()
+  {
+    name = _name;
+    disabled = _disabled;
+  }
+
+  /**
+   * @notice Indicates if this purchaser should be exposed as an option to users on the frontend.
+   * False does not necessarily mean previous subs will no longer work (see `stopped` for that).
+   */
+  function isActive() public view returns(bool)
+  {
+    return !stopped() && !disabled;
+  }
+
+  /**
+   * @notice Called by anyone to purchase or renew a key on behalf of a user.
+   * The user must have ERC-20 spending approved and the purchase must meet the terms
+   * defined during initialization.
+   */
+  function purchaseFor(
+    address payable _user,
+    address _referrer,
+    bytes memory _data
+  ) public
+    whenNotStopped()
+  {
+    uint temp = timeOfLastPurchase[_user];
+    require(isSubscription || temp == 0, 'SINGLE_USE_ONLY');
+    // `now` must be strictly larger than the timestamp of the last block
+    require(now - temp >= renewMinFrequency, 'BEFORE_MIN_FREQUENCY');
+
+    if(lock.getHasValidKey(_user))
+    {
+      temp = lock.keyExpirationTimestampFor(_user); // This reverts if getHasValidKey is false
+      require(temp <= now || temp - now <= renewWindow, 'OUTSIDE_RENEW_WINDOW');
+    }
+
+    temp = lock.keyPrice();
+    require(temp <= maxKeyPrice, 'PRICE_TOO_HIGH');
+
+    // It's okay if the lock changes tokenAddress as the ERC-20 approval is specifically
+    // the token the endUser wanted to spend
+    IERC20 token = IERC20(lock.tokenAddress());
```

Comment:
Hmm maybe we should check if the keyPrice is 0, and then avoid the transfer / approval.

Reply:
fixed